zepid/causal/doublyrobust/TMLE.py

```diff
@@ -800,15 +801,537 @@ def plot_love(self, color_unweighted='r', color_weighted='b', shape_unweighted='
+        fmt = 'Minimum value:           {:<20}'
+        print(fmt.format(np.round(np.min(w), decimals=decimal)))
+        fmt = 'Maximum value:           {:<20}'
+        print(fmt.format(np.round(np.max(w), decimals=decimal)))
+
+        print('----------------------------------------------------------------------')
+        print('                        Targeting Step                                ')
+        fmt = 'Epsilon:                 {:<20}'
+        print(fmt.format(np.round(self.epsilon, decimals=decimal)))
+
+        print('======================================================================')
+
+        # Generating plots
+        df = self.df.copy()
+        df['_g1_'] = np.where(df[self.exposure] == 1, self._denominator_, 1 - self._denominator_)
+
+        plt.figure(figsize=[8, 4])
+        plt.subplot(121)
+        plot_kde(df=df, treatment=self.exposure, probability='_g1_')
+        plt.title("Kernel Density of Propensity Scores")
+
+        plt.subplot(122)
+        plot_kde_accuracy(values=value, color='green')
+        plt.title("Kernel Density of Accuracy")
+        plt.tight_layout()
+        plt.show()
+
+    @staticmethod
+    def targeting_step(y, q_init, iptw, verbose):
+        f = sm.families.family.Binomial()
+        log = sm.GLM(y,  # Outcome / dependent variable
+                     np.repeat(1, y.shape[0]),  # Generating intercept only model
+                     offset=np.log(probability_to_odds(q_init)),  # Offset by g-formula predictions
+                     freq_weights=iptw,  # Weighted by calculated IPW
+                     family=f).fit()
+
+        if verbose:  # Optional argument to print each intermediary result
+            print('==============================================================================')
+            print('Targeting Model')
+            print(log.summary())
+
+        return log.params[0]  # Returns single-step estimated Epsilon term
+
     @staticmethod
-    def _unit_bounds(y, mini, maxi, bound):
-        # bounding for continuous outcomes
-        v = (y - mini) / (maxi - mini)
-        v = np.where(np.less(v, bound), bound, v)
-        v = np.where(np.greater(v, 1-bound), 1-bound, v)
-        return v
+    def est_marginal_variance(haw, y_obs, y_pred, y_pred_targeted, psi):
+        # https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4965321/
+        doqg_psi_sq = (haw*(y_obs - y_pred) + y_pred_targeted - psi)**2
+        var_est = np.mean(doqg_psi_sq)
+        return var_est
 
     @staticmethod
-    def _unit_unbound(ystar, mini, maxi):
-        # unbounding of bounded continuous outcomes
-        return ystar*(maxi - mini) + mini
+    def est_conditional_variance(haw, y_obs, y_pred):
+        # https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4965321/
+        doqg_psi_sq = (haw*(y_obs - y_pred))**2
+        var_est = np.mean(doqg_psi_sq)
+        return var_est
+
+
+# Functions that all TMLEs can call are below
+def _tmle_unit_bounds_(y, mini, maxi, bound):
```

Comment:
this makes me think you may want a `_TMLE` super class with (atleast) these two methods, and any other common overlap between `TMLE` and `StochasticTMLE`. It maybe premature, and premature abstractions can be bad.